server/image_response.js

```diff
@@ -12,19 +13,27 @@ module.exports = function(db) {
     module.get_img = function (req, res) {
         let id = req.query.id;
         let large = req.query.large;
-        //var match_array = new RegExp('\/images\/(.*+?)', 'g').exec(url);
-        let img_dir = dir + 'cell_images/';
-        if (large) {
-            fs.readdir(img_dir, function(err, items) {
-                res.sendFile(img_dir + items[id - 1]);
-            });
-        }
-        else {
-            // send smaller files
-            fs.readdir(img_dir + "ds/", function (err, items) {
-                res.sendFile(img_dir + "ds/" + items[id - 1]);
+        db.one("SELECT * FROM images WHERE id=$1", id)
+            .then((image) => {
+                // image found
+                let imagePath;
+                let imageFileName = image.hash + image.extension;
+                if (large) { imagePath = path.join(image.directory, imageFileName) }
+                // get downsampled image instead
+                else { imagePath = path.join(path.join(image.directory, 'ds/'), imageFileName)}
+                fs.access(imagePath, fs.constants.F_OK, (err) => {
+                    if (err) {
+                        res.status(500).send(`Could not access image ${imageFileName} in file system`);
```

Comment:
Haha sorry, in essence I went with "400 means the client messed up and 500 means the server messed up". When I start refactoring error handling, I'll address this (and also chat with you about what's appropriate).